Adds the mystery box

Ryll-Ryll opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
A compliment to my recent root beer PR, since I've been on a retro CoD Zombies kick lately, this PR adds a new admin structure that randomly rolls an item and presents it to the player like the mystery box from... well, CoD Zombies.

[Here's the new public domain version in action.](https://streamable.com/01btp8) The music that plays while selecting is The Ragtime Drummer by James Lent, sourced from [freemusicarchive](https://freemusicarchive.org/music/Antique_Phonograph_Music_Program_Various_Artists/Antique_Phonograph_Music_Program_11032009/James_Lent_-_The_Ragtime_Drummer).

There's a base version that can spawn any item that the christmas tree/santa's bag can spawn, and then there's a subtype that draws from the Summon Guns list only. If you want, you can use the base type to generate a list of any subtypes of a given path, or you can edit the valid_types list to only consider specific types of atoms.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Fun in the thunderdome, giving players a fun way to spawn in weapons.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl: Ryll/Shaps
add: Added in a new admin spawn, the mystery box! Dealing randomized (maybe) death to thunderdomes near you! Maybe!
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Ryll-Ryll committed:
initial mystery box PR

Ryll-Ryll committed:
tick

Ryll-Ryll committed:
attr

Ryll-Ryll committed:
Merge branch 'master' into press-f-for-a-random-weapon--cost-950

tralezab commented:
@Ryll-Ryll stop playing so much kino der toten i think it's starting to affect your pr ideas

Jackraxxus commented:
If this is meant for thunderdome gaming could you maybe blacklist the staff/wand of safety?
It teleports people hit to the station.

Ryll-Ryll commented:
> @Ryll-Ryll stop playing so much kino der toten i think it's starting to affect your pr ideas

pack a punch next

> If this is meant for thunderdome gaming could you maybe blacklist the staff/wand of safety? It teleports people hit to the station.

Yeah, I'll include some presets and easy tools to change them shortly

AndroidMK2 commented:
Start playing barotrauma so we get our sexy underwater map

StoneMason commented:
it should take 950 credits to use

optimumtact commented:
@Ryll-Ryll free to download doesn't necessarily translate into free to use, do you know what actual license it's under?

Ryll-Ryll committed:
avoids getting sued for copyright breach

Ryll-Ryll committed:
adds a gun subtype

Ryll-Ryll committed:
Merge remote-tracking branch 'Ryll-Ryll/press-f-for-a-random-weapon--cost-950' into press-f-for-a-random-weapon--cost-950

Ryll-Ryll commented:
Alright clearly I was misled by youtube comments about Slapstick Dick's copyright status, I've replaced it with a new public domain tune called The Ragtime Drummer by James Lent. Also added some new functionality, so it can currently act like a somewhat flashier yet slower christmas tree, as well as a gun dispenser

Ryll-Ryll committed:
Merge branch 'master' into press-f-for-a-random-weapon--cost-950

Ryll-Ryll committed:
expanding on the box types

Ryll-Ryll committed:
Merge remote-tracking branch 'Ryll-Ryll/press-f-for-a-random-weapon--cost-950' into press-f-for-a-random-weapon--cost-950

Ryll-Ryll committed:
fixes bad var name

Ryll-Ryll committed:
more content

tralezab commented on code/game/objects/structures/mystery_box.dm:
```diff
@@ -0,0 +1,278 @@
+/obj/structure/mystery_box/attack_hand(mob/living/user, list/modifiers)
+	. = ..()
+	switch(box_state)
+		if(MYSTERY_BOX_COOLING_DOWN)
+
```
remove empty switch cases, i know why they're there but they don't really help reading that much

tralezab commented on code/game/objects/structures/mystery_box.dm:
```diff
@@ -0,0 +1,278 @@
+		if(MYSTERY_BOX_STANDBY)
+			activate(user)
+
+		if(MYSTERY_BOX_CHOOSING)
+
```
see above

tralezab commented on code/game/objects/structures/mystery_box.dm:
```diff
@@ -0,0 +1,278 @@
+	valid_types = list()
+
+	for(var/iter_path in typesof(selectable_base_type))
+		if(ispath(iter_path) && ispath(iter_path, /obj/item))
```
early continue, why not?

tralezab commented on code/game/objects/structures/mystery_box.dm:
```diff
@@ -0,0 +1,278 @@
+
+	var/matrix/starting = matrix()
+	animate(src, pixel_y = 6, transform = starting, time = MBOX_DURATION_CHOOSING, easing = QUAD_EASING | EASE_OUT)
+	loop_icon_changes()
+
+/// Keep changing the icon and selected path
+/obj/mystery_box_item/proc/loop_icon_changes()
+	set waitfor = FALSE
```
INVOKE_ASYNC the proc instead

LemonInTheDark commented on code/game/objects/structures/mystery_box.dm:
```diff
@@ -0,0 +1,278 @@
+	/// The uninstantiated item that's currently selected based off selected_path, for use with initial()
+	var/obj/item/selected_item
+
+	while(world.time < start_time + MBOX_DURATION_CHOOSING)
+		change_delay += change_delay_delta
+		selected_path = pick(parent_box.valid_types)
+		selected_item = selected_path
+		icon = initial(selected_item.icon)
+		icon_state = initial(selected_item.icon_state)
+		sleep(change_delay)
+
+	add_filter("ready_outline", 2, list("type" = "outline", "color" = "#FBFF23", "size" = 0.2))
+	name = initial(selected_item.name)
+	parent_box.present_weapon()
+	claimable = TRUE
```
I think these could be a series of animates and a timer, would make life a bit easier

LemonInTheDark replied:
Wow it just doesn't work. that's depressing

Ryll-Ryll replied:
Yeah I was messing with it over the last few days with different configurations, but it appears to just take the last one and run it immediately. I'm not thrilled about using all timers instead but meh

Ryll-Ryll committed:
blorble broke the loop

Ryll-Ryll committed:
blorble

Ryll-Ryll committed:
makes the icon switching work

Ryll-Ryll committed:
removes debug vars

Ryll-Ryll committed:
removes one more reference to an unused var

Ryll-Ryll committed:
per arm's reviews

Ryll-Ryll committed:
one more review change

cacogen commented:
bioshock vibes
